src/main/kotlin/com/exactpro/th2/ws/client/api/impl/WebSocketClient.kt

```diff
@@ -38,40 +40,40 @@ class WebSocketClient(
     private val logger = KotlinLogging.logger {}
     private val textFrames = mutableListOf<String>()
     private val binaryFrames = mutableListOf<ByteArray>()
-    @Volatile private lateinit var socket: WebSocket
+    private val lock = ReentrantLock()
+    private lateinit var socket: WebSocket
 
     @Volatile var isRunning: Boolean = false
         private set
 
     private fun awaitSocket(): WebSocket {
-        logger.debug { "Waiting for a connected socket" }
-
-        while (!::socket.isInitialized || socket.isOutputClosed) {
-            Thread.sleep(1)
+        when {
+            !isRunning -> start()
+            ::socket.isInitialized || socket.isOutputClosed -> connect()
         }
 
-        logger.debug { "Acquired connected socket" }
-
         return socket
     }
 
-    override fun sendText(text: String) {
+    override fun sendText(text: String) = lock.withLock {
         logger.debug { "Sending text: $text" }
         val preparedText = handler.prepareText(this, text)
         awaitSocket().sendText(preparedText, true)
         onMessage(preparedText.toByteArray(), true, SECOND)
     }
 
-    override fun sendBinary(data: ByteArray) {
+    override fun sendBinary(data: ByteArray) = lock.withLock {
         logger.debug { "Sending binary: ${data.toBase64()}" }
         val preparedData = handler.prepareBinary(this, data)
         awaitSocket().sendBinary(ByteBuffer.wrap(preparedData), true)
         onMessage(preparedData, false, SECOND)
     }
 
     override fun sendPing(message: ByteArray) {
-        logger.debug { "Sending ping: ${message.toBase64()}" }
-        awaitSocket().sendPing(ByteBuffer.wrap(message))
+        lock.withLock {
```

Comment:
override fun sendPing(message: ByteArray) = lock.withLock {

Reply:
Would work only if we explicitly specify `sendPing`'s return type as a `Unit`. Otherwise it would fail because `lock.withLock` here returns a `CompletableFuture`